### PR TITLE
coverage with clang

### PR DESCRIPTION
**Pull request by nilsthamm**
#905 
Enables clang coverage on Linux (to test JIT files). Coverage diff will still be executed with gcc (and only on linux), since llvm does not have a usable export file type (https://llvm.org/docs/CommandGuide/llvm-cov.html#llvm-cov-report)

### CODE REVIEW

**Comment by mrks** [comment]
Have you seen https://git.phoenix.dj/phoenix/llvm-cov-xml ? Not sure if that helps.

File: scripts/coverage.sh [hunk 1]
```diff
@@ -22,80 +22,80 @@ fi
...
-if [[ "$unamestr" == 'Darwin' ]]; then
-  # LLVM has its own way of dealing with coverage...
-  # https://llvm.org/docs/CoverageMappingFormat.html
+# LLVM has its own way of dealing with coverage...
+# https://llvm.org/docs/CoverageMappingFormat.html
   
-  # merge the profile data using the llvm-profdata tool:
-  /usr/local/opt/llvm/bin/llvm-profdata merge -o ./default.profdata ./default.profraw
+# merge the profile data using the llvm-profdata tool:
+$path_to_compiler'llvm-profdata' merge -o ./default.profdata ./default.profraw
```
Comment (lawben) [inline scripts/coverage.sh]: i think this is easier to read `"${path_to_compiler}llvm-profdata"`. not that is especially nice to read but i was really confused about the single quotes

File: scripts/coverage.sh [hunk 2]
```diff
@@ -22,80 +22,80 @@ fi
...
 
-  # run LLVM’s code coverage tool
-  /usr/local/opt/llvm/bin/llvm-cov show -format=html -instr-profile ./default.profdata build-coverage/hyriseTest -output-dir=./coverage ./src/lib/
+# run LLVM’s code coverage tool
+$path_to_compiler'llvm-cov' show -format=html -instr-profile ./default.profdata build-coverage/hyriseTest -output-dir=./coverage ./src/lib/
```
Comment (lawben) [inline scripts/coverage.sh]: same as above

File: scripts/coverage.sh [hunk 3]
```diff
@@ -22,80 +22,80 @@ fi
...
 
-  rm default.profdata default.profraw
+# rm default.profdata default.profraw
```
Comment (lawben) [inline scripts/coverage.sh]: is this needed or can the line be removed?

File: scripts/coverage.sh [hunk 4]
```diff
@@ -22,80 +22,60 @@ fi
...
-   cmake -DCMAKE_CXX_COMPILER_LAUNCHER=$launcher -DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DENABLE_COVERAGE=ON ..
-elif [[ "$unamestr" == 'Darwin' ]]; then
-   # Use Clang for OS X
-   cmake -DCMAKE_CXX_COMPILER_LAUNCHER=$launcher -DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_COMPILER=/usr/local/opt/llvm/bin/clang -DCMAKE_CXX_COMPILER=/usr/local/opt/llvm/bin/clang++ -DENABLE_COVERAGE=ON ..
+if [[ "$unamestr" == 'Darwin' ]]; then
+   # Use homebrew clang for OS X
+   path_to_compiler='/usr/local/opt/llvm/bin/'
 fi
 
+cmake -DCMAKE_CXX_COMPILER_LAUNCHER=$launcher -DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_COMPILER=$path_to_compiler'clang' -DCMAKE_CXX_COMPILER=$path_to_compiler'clang++' -DENABLE_COVERAGE=ON ..
```
Comment (mrks) [inline scripts/coverage.sh]: use `${compiler_path}/clang`

File: scripts/coverage.sh [hunk 5]
```diff
@@ -22,80 +22,60 @@ fi
...
 cd -
 
 rm -fr coverage; mkdir coverage
 ./build-coverage/hyriseTest build-coverage --gtest_filter=-SQLiteTestRunnerInstances/*
 
-if [[ "$unamestr" == 'Darwin' ]]; then
-  # LLVM has its own way of dealing with coverage...
-  # https://llvm.org/docs/CoverageMappingFormat.html
+# LLVM has its own way of dealing with coverage...
+# https://llvm.org/docs/CoverageMappingFormat.html
```
Comment (mrks) [inline scripts/coverage.sh]: This comment doesn't make sense anymore, does it?

File: scripts/coverage.sh [hunk 6]
```diff
@@ -22,80 +22,60 @@ fi
...
+
+# run LLVM’s code coverage tool
+$path_to_compiler'llvm-cov' show -format=html -instr-profile ./default.profdata build-coverage/hyriseTest -output-dir=./coverage ./src/lib/
+
+echo Coverage Information is in ./coverage/index.html
+
+# Continuing only if diff output is needed with Linux/gcc
+if [ "true" == "$generate_badge" ]; then
+
+  $path_to_compiler'llvm-cov' export -summary-only -instr-profile ./default.profdata build-coverage/hyriseTest ./src/lib/ > coverage.json
```
Comment (mrks) [inline scripts/coverage.sh]: see above

**Comment by mrks** [comment]
"Coverage — decreased by 17.35 compared to master". Saw this a bit late. Any idea what happened there?

In any case, we won't get around using clang for coverage if we want to have a look at JIT.

**Comment by nilsthamm** [comment]
First I thought JIT is the problem here that it is comparably bad covered. But I'm not so sure anymore. It could be the export command doesn't recognize the directory we want to get the coverage information for. I will have a look at it...

**Pull request merged**